### ISSUE TITLE
XLA FusedBatchNormV3 produces negative variance and NaN output

### ISSUE THREAD

**Issue by wuyii8941** [tensorflow/tensorflow#118701]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.22.0-dev20260429

### Custom code

Yes

## Summary

`FusedBatchNormV3` compiled with `jit_compile=True` (is_training=True) produces NaN output when input data has a large constant offset. The root cause is XLA's BatchNormExpander pass using the numerically unstable one-pass variance formula `E[x²] - E[x]²`, which suffers from catastrophic cancellation and computes a **negative variance**. Taking sqrt of this negative value produces NaN. The gradient is also completely broken (all NaN), at an even lower threshold.

The bug manifests at surprisingly low offsets depending on configuration:
- **NHWC, 1 channel**: NaN at offset ≥ 1e7
- **NCHW, 3 channels**: NaN at offset ≥ **1e4** (10,000)
- **NHWC, 8+ channels**: NaN at offset ≥ **1e5** (100,000)
- **Gradient**: NaN at offset ≥ **1e6** (even when forward pass is clean)

Eager mode uses cuDNN's numerically stable fused batch norm kernel and produces correct results.

## Root cause

XLA's `BatchNormExpander` pass decomposes FusedBatchNormV3 into primitive operations. For computing the batch variance, it uses the one-pass formula:

```
var = E[x²] - E[x]²
```

When the mean is large (e.g., 1e8), both `E[x²]` and `E[x]²` are approximately `1e16`, but float32 only has ~7 digits of precision. The subtraction cancels nearly all significant digits, and the rounding errors can make the result negative (an impossible value for variance). Taking `sqrt(negative_var)` produces NaN.

cuDNN's fused batch norm kernel uses Welford's online algorithm or a two-pass algorithm that computes `E[(x - mean)²]`, which avoids this catastrophic cancellation entirely.

## Variance progression with increasing offset (NHWC, 1 channel)

| Offset | Eager var | XLA var | XLA NaN? |
|--------|-----------|---------|----------|
| 1e4 | 1.01 | 0.00 | No |
| 1e5 | 0.87 | 1028.0 | No |
| 1e6 | 1.04 | 0.00 | No |
| 1e7 | 1.07 | -8.4e6 | **YES** |
| 1e8 | 0.00 | -1.08e9 | **YES** |

Note: even at 1e5, the XLA variance (1028.0) is already dramatically wrong compared to the true variance (~0.87). The bug is present at lower offsets as a precision issue and becomes catastrophic when variance goes negative.

## NCHW format breaks at offset 1e4

With NCHW data format, the bug manifests at a much lower offset:

| Offset | NCHW NaN? | XLA variance (3 channels) |
|--------|-----------|---------------------------|
| 1e3 | No | [0.88, 1.00, 1.13] (correct) |
| **1e4** | **YES** (1/3 ch) | [16.06, 0.00, **-8.03**] |
| 3e4 | **YES** (1/3 ch) | [0.00, **-128.5**, 0.00] |
| 1e5 | **YES** (1/3 ch) | [0.00, 0.00, **-1028**] |
| 1e6 | **YES** (1/3 ch) | [131586, 65793, **-65793**] |
| 3e6 | **YES** (all) | all negative |

## Multi-channel makes it worse

With more channels, statistical variation means some channels go negative at lower offsets:

| Channels | Lowest offset with NaN | Channels affected |
|----------|----------------------|-------------------|
| 1 | 1e7 | 1/1 |
| 3 | 1e6 | 1/3 |
| 8 | **1e5** | 3/8 |
| 16 | **1e5** | 7/16 |
| 32 | **1e5** | 11/32 |

## Gradient is broken at LOWER threshold (1e6)

The gradient (`FusedBatchNormGradV3`) breaks at an even lower offset than the forward pass:

| Offset | Forward NaN? | Gradient NaN? |
|--------|-------------|---------------|
| 1e5 | No | No |
| 1e6 | No | **YES** (512/512 NaN) |
| 1e7 | No | **YES** (2048/2048 NaN) |
| 1e8 | **YES** | **YES** |

The gradient breaks at offset=1e6 while the forward pass only produces NaN at 1e8. This means **training is silently corrupted at a much lower threshold** than where visible NaN output appears in the forward pass.

```python
@tf.function(jit_compile=True)
def fbn_grad_xla(x):
    with tf.GradientTape() as tape:
        tape.watch(x)
        r = tf.raw_ops.FusedBatchNormV3(
            x=x, scale=scale, offset=offset, mean=mean_init, variance=var_init,
            is_training=True, data_format='NHWC')
        loss = tf.reduce_sum(r.y)
    return tape.gradient(loss, x)

x = tf.constant(np.random.randn(4,8,8,1).astype(np.float32) + 1e6)
grad = fbn_grad_xla(x)  # All NaN!
```

## Additional findings

- **Inference mode** (is_training=False): Not affected (uses provided mean/variance)
- **Grappler** (jit_compile=False): Not affected (uses cuDNN kernel)
- **tf.math.reduce_variance under XLA**: Not affected (uses the correct two-pass formula)
- **Group normalization under XLA**: Also shows precision loss at large offsets (93% relative error at 1e8), likely same root cause
- This is specific to `FusedBatchNormV3`'s XLA decomposition path

## Impact

This affects any model using batch normalization under XLA (`jit_compile=True`) when input features have large constant offsets. An offset of just **10,000** (1e4) is sufficient to trigger NaN with NCHW format, and **100,000** (1e5) with multiple channels in NHWC. Common scenarios:
- Sensor data (absolute temperature in millikelvin, pressure in Pa)
- Financial data (stock prices, timestamps, account balances)
- Features that haven't been centered before batch norm
- Pixel values in unnormalized images (0-65535 for 16-bit images)

## Environment

- TF: 2.22.0-dev20260429
- GPU: NVIDIA (CUDA)

### Standalone code to reproduce the issue

```shell
## Reproducer


import tensorflow as tf
import numpy as np

with tf.device('/GPU:0'):
    np.random.seed(42)
    x = np.random.randn(4, 8, 8, 1).astype(np.float32) + 1e8
    x_t = tf.constant(x)
    scale = np.ones(1, dtype=np.float32)
    offset = np.zeros(1, dtype=np.float32)
    mean_init = np.zeros(1, dtype=np.float32)
    var_init = np.ones(1, dtype=np.float32)

    # Eager: no NaN
    eager_r = tf.raw_ops.FusedBatchNormV3(
        x=x_t, scale=scale, offset=offset, mean=mean_init, variance=var_init,
        is_training=True, data_format='NHWC')
    print(f"Eager NaN: {np.isnan(eager_r.y.numpy()).sum()}")  # 0

    # XLA: all NaN due to negative variance
    @tf.function(jit_compile=True)
    def fbn_xla(x):
        return tf.raw_ops.FusedBatchNormV3(
            x=x, scale=scale, offset=offset, mean=mean_init, variance=var_init,
            is_training=True, data_format='NHWC')
    xla_r = fbn_xla(x_t)
    print(f"XLA NaN: {np.isnan(xla_r.y.numpy()).sum()}")  # 256
    print(f"XLA variance: {xla_r.batch_variance.numpy()}")  # [-1.0779526e+09] (NEGATIVE!)
```

### Relevant log output

```shell

```

**Comment by dhantule** [comment]
Hi @wuyii8941, 
Thanks for reporting this and sharing the details.
We’ve received the issue and are taking a look. We’ll follow up here once we have more information or if we need any additional inputs.
Appreciate your patience.
